consensus/istanbul/backend/kip103.go

```diff
@@ -0,0 +1,179 @@
+
+	context := blockchain.NewEVMContext(msg, caller.header, caller.chain, nil)
+	evm := vm.NewEVM(context, caller.state, caller.chain.Config(), &vm.Config{}) // no additional vm config required
+
+	res, _, kerr := blockchain.ApplyMessage(evm, msg)
+	return res, kerr.ErrTxInvalid
+}
+
+type kip103Receipt struct {
+	Retired map[common.Address]*big.Int `json:"retired"`
+	Newbie  map[common.Address]*big.Int `json:"newbie"`
+	Burnt   *big.Int                    `json:"burnt"`
+	Success bool                        `json:"success"`
+}
+
+func newKip103Receipt() *kip103Receipt {
+	return &kip103Receipt{
+		Retired: make(map[common.Address]*big.Int),
+		Newbie:  make(map[common.Address]*big.Int),
+		Burnt:   big.NewInt(0),
+		Success: false,
+	}
+}
+
+func (receipt *kip103Receipt) fillRetired(contract *kip103.TreasuryRebalanceCaller, state *state.StateDB) error {
+	numRetiredBigInt, err := contract.GetRetiredCount(nil)
+	if err != nil {
+		logger.Error("Failed to get RetiredCount from TreasuryRebalance contract", "err", err)
+		return err
+	}
+
+	for i := 0; i < int(numRetiredBigInt.Int64()); i++ {
+		ret, err := contract.Retirees(nil, big.NewInt(int64(i)))
+		if err != nil {
+			logger.Error("Failed to get Retirees from TreasuryRebalance contract", "err", err)
+			return err
+		}
+		receipt.Retired[ret] = state.GetBalance(ret)
+	}
+	return nil
+}
+
+func (receipt *kip103Receipt) fillNewbie(contract *kip103.TreasuryRebalanceCaller) error {
+	numNewbieBigInt, err := contract.GetNewbieCount(nil)
+	if err != nil {
+		logger.Error("Failed to get NewbieCount from TreasuryRebalance contract", "err", err)
+		return nil
+	}
+
+	for i := 0; i < int(numNewbieBigInt.Int64()); i++ {
+		ret, err := contract.Newbies(nil, big.NewInt(int64(i)))
+		if err != nil {
+			logger.Error("Failed to get Newbies from TreasuryRebalance contract", "err", err)
+			return err
+		}
+		receipt.Newbie[ret.Newbie] = ret.Amount
+	}
+	return nil
+}
+
+func (receipt *kip103Receipt) totalRetriedBalance() *big.Int {
+	total := big.NewInt(0)
+	for _, bal := range receipt.Retired {
+		total.Add(total, bal)
+	}
+	return total
+}
+
+func (receipt *kip103Receipt) totalNewbieBalance() *big.Int {
+	total := big.NewInt(0)
+	for _, bal := range receipt.Newbie {
+		total.Add(total, bal)
+	}
+	return total
+}
+
+// rebalanceTreasury reads data from a contract, validates stored values, and executes treasury rebalancing (KIP-103).
+// It can change the global state by removing old treasury balances and allocating new treasury balances.
+// The new allocation can be larger than the removed amount, and the difference between two amounts will be burnt.
+func rebalanceTreasury(state *state.StateDB, chain consensus.ChainReader, header *types.Header) (*kip103Receipt, error) {
+	receipt := newKip103Receipt()
+	c := &kip103ContractCaller{state, chain, header}
+
+	// Inside check to avoid a panic case
+	if chain.Config().KIP103 == nil {
+		return receipt, errors.New("cannot find KIP103 configuration")
```

Comment:
```suggestion
		return nil, errors.New("cannot find KIP103 configuration")
```

Reply:
I'm sorry, but i will skip this comment. Maybe we can refine this later